clients/include/testing_copy_batched.hpp

```diff
@@ -60,7 +60,7 @@ void testing_copy_batched(const Arguments& arg)
             return;
         }
 
-        if(batch_count < 0)
+        if(N > 0 && batch_count < 0)
```

Comment:
Not sure if I like an invalid_parameter when N >= 0 but okay when N <= 0.
I would think the invalid parameter should take precedence.

Reply:
> Not sure if I like an invalid_parameter when N >= 0 but okay when N <= 0.
> I would think the invalid parameter should take precedence.

(Actually, in `SCAL`, `batch_count < 0` is an invalid argument only if `N > 0`. `batch_count < 0 && (N <= 0 || incx <= 0)` is not an error.)

As explained in [SWDEV-203994](http://ontrack-internal.amd.com/browse/SWDEV-203994), if a kernel has enough information to know the result (including that it's a no-op), such as if `N <= 0`, then it does not matter what the other arguments are, and it's not an error.

If `N <= 0` (or `incx <= 0` in the case of `SCAL`), then it does not matter if `batch_count < 0`, because the result is known to be a no-op in `SCAL`, `AXPY`, `SWAP`, etc.

For `DOT`, `ASUM`, `NRM2`, etc., where `N <= 0` implies a _constant_ result, then `batch_count < 0` is an error, because there is no way to return _a negative number of results_.

When `N <= 0`, it does not matter whether `x == nullptr`, `y == nullptr`, `alpha == nullptr`, etc., because those parameters are not going to affect the result of either a no-op (e.g., `SCAL`, `AXPY`, `SWAP`) or a constant return value (e.g., `DOT`, `ASUM`, `NRM2`). 

BLAS was written in "Fortran philosophy": If you have enough information to determine the answer, then that's it, and you don't have to validate all arguments.

@mahmoodw @TorreZuk @amcamd @YvanMokwinski @daineAMD

Reply:
This guide should also be put in the doxygen general usage information on Blas or blas1.   Anyone who has never used fortran (most programmers) would not expect such rules.

Reply:
> This guide should also be put in the doxygen general usage information on Blas or blas1. Anyone who has never used fortran (most programmers) would not expect such rules.

It is CuBLAS behavior.

It is not that Fortran-specific, but is **mathematical**, similar to `x * 0 == 0` (`NaN` and `Inf` notwithstanding). If you know the answer because of a mathematical identity, then you do not need to validate all arguments, since the identity still holds.

If you are asked to multiply a 0-sized matrix or vector by a scalar pointed to by a pointer, then you do not need to dereference that pointer, and thus you do not need to validate it, since the 0-sized matrix is empty, and the operation is a no-op.

It is similar to [lazy evaluation](https://en.wikipedia.org/wiki/Lazy_evaluation), specifically,  [avoiding error conditions](https://en.wikipedia.org/wiki/Lazy_evaluation#Avoidance_of_error_conditions).

Lazy evaluation is common in functional languages, and mathematical languages like R.

See [SWDEV-203994](http://ontrack-internal.amd.com/browse/SWDEV-203994) and [PR 722](https://github.com/ROCmSoftwarePlatform/rocBLAS/pull/722).